Bb chapter 9 (#707)
mlr-org/mlr3book: book/chapters/chapter9/preprocessing.qmd

`r index('Data cleaning', aside = TRUE)`\index{exploratory data analysis|see{data cleaning}} is an important part of preprocessing that involves the removal of errors, noise, and redundancy in the data; we only consider data cleaning very briefly as it is usually performed outside of `mlr3` on the raw dataset.

Another aspect of preprocessing is `r index('feature engineering', aside = TRUE)`, which covers all other transformations of data before it is fed to the machine learning model, including the creation of features from possibly unstructured data, such as written text, sequences or images.
The goal of feature engineering is to prepare the data so that a model can be trained on it, and/or to further improve predictive performance.
It is important to note that feature engineering helps mostly for simpler algorithms, while highly complex models usually gain less from it and require little data preparation to be trained.

> Common difficulties in data that can be solved with feature engineering include features with (high) skew distributions, high cardinality categorical features, missing observations, high dimensionality and imbalanced classes in classification tasks.

Reviewer: @mb706 @berndbischl  you've removed all definition of feature engineering? Now this paragraph is hard to follow

Author: I think this part got lost when transferring to gdoc and back, I will add it again

> Hence, manual feature engineering is still often required but with `mlr3pipelines`, we can simplify the process as much as possible.

Reviewer: ```suggestion
Hence, manual feature engineering is still often required but with `mlr3pipelines`, which can simplify the process as much as possible.
```

`r index('Data cleaning', aside = TRUE)`\index{exploratory data analysis|see{data cleaning}} is an important part of preprocessing that involves the removal of errors, noise, and redundancy in the data; we only consider data cleaning very briefly as it is usually performed outside of `mlr3` on the raw dataset.

Another aspect of preprocessing is `r index('feature engineering', aside = TRUE)`, which covers all other transformations of data before it is fed to the machine learning model, including the creation of features from possibly unstructured data, such as written text, sequences or images.

> The goal of feature engineering is to bring the data in a format that a given learner can handle, and/or to further improve predictive performance.

Reviewer: ```suggestion
The goal of feature engineering is to enable the data to be handled by a given learner, and/or to further improve predictive performance.
```

2. Compute the mean of the target variable for each group.
3. Compute the global mean of the target variable.
4. Compute the impact score for each group as the difference between the mean of the target variable for the group and the global mean of the target variable.
5. Replace the categorical feature with the impact scores.

Impact encoding preserves the information of the categorical feature while also creating a numerical representation that reflects its importance in predicting the target.
Compared to one-hot encoding, the main advantage is that only a single numeric feature is created regardless of the number of levels of the categorical features, hence it is especially useful for high-cardinality features.
As information from the target is used to compute the impact scores, the encoding process must be embedded in cross-validation to avoid leakage between training and testing data (@sec-performance).

As well as encoding features, other basic preprocessing steps for categorical features include removing constant features (which only have one level and may have been removed as part of data cleaning), and collapsing levels that occur very rarely.

> Unfortunately, these types of problems can occur as artifacts in resampling as dataset size is further reduced here.

Reviewer: ```suggestion
These types of problems can occur as artifacts of resampling as the dataset size is further reduced.
```

bmr$aggregate(measure = msr_mae)[, .(learner_id, regr.mae)]
```

In this small experiment, we see that the difference between the extended factor encoding pipeline and the simpler one-hot encoding strategy pipeline is only very small.
If you are interested in learning more about different encoding strategies, including a benchmark study comparing them, we recommend @pargent2022regularized.

## Missing Values {#sec-preprocessing-missing}

> A common problem in real-world data is `r index('missing values')` in features.

Reviewer: ```suggestion
A common problem in real-world data is `r index('missing values', 'missing data')` in features.
```

The pipeline includes the following `r ref("PipeOp")`s (some are applied multiple times and most use selectors):

1. `po("removeconstants")` -- Constant features are removed.
2. `po("colapply")` -- Character and ordinal features are encoded as categorical, and date/time features are encoded as numeric.
3. `po("imputehist")` -- Numeric features are imputed by histogram sampling.
4. `po("imputesample")` -- Logical features are imputed by sampling from the empirical distribution -- this only affects the `$predict()`-step.
5. `po("missind")` -- Missing data indicators are added for imputed numeric and logical variables.
6. `po("imputeoor")` -- Missing values of categorical features are encoded with a new level.
7. `po("fixfactors")` -- Fixes levels of categorical features such that the same levels are present during prediction and training (which may involve dropping empty factor levels).
8. `po("imputesample")` -- Missing values in categorical features introduced from dropping levels in the previous step are imputed by sampling from the empirical distributions.

> 9. `po("collapsefactors")` -- Categorical features levels are collapsed (starting from the rarest factors in the training data) until there are less than a certan number of levels.
>   This number is controlled by the `max_cardinality` argument and defaults to a conservative 1000.

Reviewer: ```suggestion
9. `po("collapsefactors")` -- Categorical features levels are collapsed (starting from the rarest factors in the training data) until there are less than a certan number of levels, controlled by the `max_cardinality` argument (with a conservative default of `1000`).
```

10. `po("encode")` -- Categorical features are one-hot encoded.
11. `po("removeconstants")` -- Constant features that might have been created in the previous steps are removed.

`ppl("robustify")` has optional arguments `task` and `learner`.

> If these are provided, then the resulting pipeline will be set up to handle the given task and learner specifically: it will e.g. not impute missing values if the learner has the `"missings"` property, or if there are no missing values in the task to begin with.

Reviewer: ```suggestion
If these are provided, then the resulting pipeline will be set up to handle the given task and learner specifically, for example, it will not impute missing values if the learner has the `"missings"` property, or if there are no missing values in the task to begin with.
```

These outputs look sensible compared to @fig-energy so we can now run our final benchmark experiment using feature extraction.
We do not need to add the `PipeOp` to each learner as we can apply it once (as above) before any model training by applying it to all available data.

```{r preprocessing-027, warning=FALSE, R.options = list(datatable.print.nrows = 13, datatable.print.class = FALSE, datatable.print.keys = FALSE, datatable.print.trunc.cols = TRUE)}
learners = list(lrn_baseline, lrn("regr.rpart"), glrn_xgb_impact,
    glrn_rf_impute_oor, glrn_lm_robust, glrn_log_lm_robust)

bmr_final = benchmark(benchmark_grid(c(tsk_ames_ext, tsk_ames), learners,
  rsmp_cv3))

perf = bmr_final$aggregate(measure = msr_mae)
perf[order(learner_id, task_id), .(task_id, learner_id, regr.mae)]
```

> 

Reviewer: ```suggestion
<!-- FIXME: RS READ -->
```
